### PR TITLE
expression, executor: rewrite built-in bit_length func.

### PR DESCRIPTION
**Pull request by dcadevil**

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -1243,31 +1243,26 @@ type bitLengthFunctionClass struct {
...
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinBitLengthSig) evalInt(row []types.Datum) (int64, bool, error) {
+	val, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
 	}
-	d.SetInt64(int64(len(str) * 8))
-	return d, nil
+	return int64(len(val) * 8), false, nil
```
Comment (zz-jason) [inline expression/builtin_string.go]: when input is a `binary(m)`, bit_length will return `8*m`

```sql
DROP TABLE IF EXISTS table_binary;
CREATE TABLE table_binary (
    a BINARY(20),
    b VARBINARY(20),
    c TINYBLOB,
    d BLOB(20),
    e MEDIUMBLOB,
    f LONGBLOB,
    g BIGINT
);
INSERT INTO table_binary (a,b,c,d, e, f, g) VALUES ("a", "b", "c", "d", "e", "f", 2);

select bit_length(a) from table_binary;
+---------------+
| bit_length(a) |
+---------------+
|           160 |
+---------------+
```

File: executor/executor_test.go [hunk 2]
```diff
@@ -911,6 +911,13 @@ func (s *testSuite) TestStringBuiltin(c *C) {
 	tk.MustExec(`insert into t values(1, 1.1, "2017-01-01 12:01:01", "12:01:01", "abcdef", 0b10101, "512", "abc")`)
 	result = tk.MustQuery("select to_base64(a), to_base64(b), to_base64(c), to_base64(d), to_base64(e), to_base64(f), to_base64(g), to_base64(h), to_base64(null) from t")
 	result.Check(testkit.Rows("MQ== MS4x MjAxNy0wMS0wMSAxMjowMTowMQ== MTI6MDE6MDE= YWJjZGVm ABU= NTEyAAAAAAAAAAAAAAAAAAAAAAA= YWJj <nil>"))
+
+	// for bit_length
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a int, b double, c datetime, d time, e char(20), f bit(10))")
```
Comment (zz-jason) [inline executor/executor_test.go]: add test for binary types

File: expression/builtin_string.go [hunk 3]
```diff
@@ -1279,31 +1279,32 @@ type bitLengthFunctionClass struct {
 }
 
 func (c *bitLengthFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinBitLengthSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinBitLengthSig{baseIntBuiltinFunc{bf}}
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: set bf.tp.Flen

File: expression/builtin_string.go [hunk 4]
```diff
@@ -1279,31 +1279,32 @@ type bitLengthFunctionClass struct {
...
+	val, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
 	}
 
-	str, err := args[0].ToString()
-	if err != nil {
-		return d, errors.Trace(err)
+	ft := b.args[0].GetType()
+	if mysql.HasBinaryFlag(ft.Flag) && ft.Tp == mysql.TypeString {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: If we do not check this,
what will TiDB get if it's MySQL type `BINARY`?

Comment (dcadevil) [inline expression/builtin_string.go]: For example,  A has type BINARY(20) and value is "a", do bit_length(A) will get 160 on MySQL, but TiDB get 8 without this check.
@XuHuaiyu

File: expression/builtin_string.go (hunk 3 above)
Comment (dcadevil) [inline expression/builtin_string.go]: fixed.
@XuHuaiyu

File: expression/builtin_string.go (hunk 4 above)
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ok, but I gets the right result using your branch.
Please test it again. @dcadevil

Comment (dcadevil) [inline expression/builtin_string.go]: I'm not exactly understand your said, what is the right result？Are you get 160 without check using my branch？@XuHaiyu

Comment (XuHuaiyu) [inline expression/builtin_string.go]: yes

Comment (dcadevil) [inline expression/builtin_string.go]: fixed.
@XuHuaiyu

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@dcadevil Thanks!

**Pull request merged**